test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java

```diff
@@ -0,0 +1,1226 @@
+     * </ul>
+     *
+     * @param testClass the class to be tested by the framework.
+     * @param scenarios scenarios which specify specific flags for the test VM.
+     *
+     * @see #runWithScenarios(Scenario...)
+     */
+    public static void runWithScenarios(Class<?> testClass, Scenario... scenarios) {
+        TestFramework framework = new TestFramework(testClass);
+        framework.addScenarios(scenarios);
+        framework.start();
+    }
+
+    /**
+     * Add VM flags to be used for the test VM. These flags override any VM or Javaoptions set by JTreg by default.<p>
+     * Use {@code -DPreferCommandLineFlags=true} if you want to prefer the VM or Javaoptions over the scenario flags.
+     *
+     * <p>
+     * The testing can be started by invoking {@link #start()}
+     *
+     * @param flags VM options to be applied to the test VM.
+     * @return the same framework instance.
+     */
+    public TestFramework addFlags(String... flags) {
+        TestRun.check(flags != null && Arrays.stream(flags).noneMatch(Objects::isNull), "A flag cannot be null");
+        this.flags.addAll(Arrays.asList(flags));
+        return this;
+    }
+
+    /**
+     * Add helper classes that can specify additional compile command annotations ({@link ForceCompile @ForceCompile},
+     * {@link DontCompile @DontCompile}, {@link ForceInline @ForceInline}, {@link DontInline @DontInline}) to be applied
+     * while testing {@code testClass} (also see description of {@link TestFramework}).
+     * <ul>
+     *     <li><p>If a helper class is not in the same file as the test class, make sure that JTreg compiles it by using
+     *            {@code @compile} in the JTreg header comment block.</li>
+     *     <li><p>If a class is used by the test class that does not specify any compile command annotations, you do not
+     *            need to include it with this method. If no helper class specifies any compile commands, you do
+     *            not need to call this method at all.</li>
+     * </ul>
+     *
+     * <p>
+     * The testing can be started by invoking {@link #start()}
+     *
+     * @param helperClasses helper classes containing compile command annotations ({@link ForceCompile},
+     *                      {@link DontCompile}, {@link ForceInline}, {@link DontInline}) to be applied
+     *                      while testing {@code testClass} (also see description of {@link TestFramework}).
+     * @return the same framework instance.
+     */
+    public TestFramework addHelperClasses(Class<?>... helperClasses) {
+        TestRun.check(helperClasses != null && Arrays.stream(helperClasses).noneMatch(Objects::isNull), "A Helper class cannot be null");
+        if (this.helperClasses == null) {
+            this.helperClasses = new ArrayList<>();
+        }
+
+        for (Class<?> helperClass : helperClasses) {
+            TestRun.check(!this.helperClasses.contains(helperClass), "Cannot add the same class twice: " + helperClass);
```

Comment:
won't it be easy to use `Set` to store `helperClasses`?

Reply:
you might also want to update javadoc for this method to state that there can be duplicates.

Reply:
you don't have to ignore duplicates, you can still throw an exception if you encounter one. I, personally, don't think we should force users to avoid dups (as I don't yet? see any issues w/ duplicated helpers), but if you think we need to enforce that *and* make it users' responsibilities then you should continue to throw an exception. in both cases, javadoc should be updated to reflect that duplicates are ignored or that an exception is raised in case a duplicate was encountered.